zebra-state/src/service/finalized_state.rs

```diff
@@ -275,6 +269,27 @@ impl FinalizedState {
         let mut orchard_note_commitment_tree = self.orchard_note_commitment_tree();
         let mut history_tree = self.history_tree();
 
+        // Check the block commitment. For Nu5-onward, the block hash commits only
+        // to non-authorizing data (see ZIP-244). This checks the authorizing data
+        // commitment, making sure the entire block contents were commited to.
+        // The test is done here (and not during semantic validation) because it needs
+        // the history tree root. While it _is_ checked during contextual validation,
+        // that is not called by the checkpoint verifier, and keeping a history tree there
+        // would be harder to implement.
+        check::finalized_block_commitment_is_valid_for_chain_history(
+            &finalized,
+            self.network,
+            &history_tree,
+        )?;
```

Comment:
I was a bit surprised to see us validating the history commitment before updating the history tree.
That means we're using the authorizing data root for the current block's transactions,
but the chain history as of the last transaction in the previous block.
After re-reading the ZIPs, I realised it was correct.

Do you think we should mention it in a comment?
Maybe it belongs in `block_commitment_is_valid_for_chain_history`?

Reply:
I would mention it yeah

Reply:
I did a double take when writing that too :laughing: I'll add a comment